client/gutenberg/extensions/videopress/edit.js

```diff
@@ -0,0 +1,39 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { Component } from '@wordpress/element';
+import apiFetch from '@wordpress/api-fetch';
+
+const withVideoPressEdit = createHigherOrderComponent( CoreVideoEdit => {
+	return class VideoPressEdit extends Component {
+		setAttributes = async attributes => {
+			const { id } = attributes;
+
+			if ( id ) {
+				const media = await apiFetch( {
+					path: `/rest/v1.1/media/${ id }`,
+				} );
+				if ( media.videopress_guid ) {
+					const videoPressData = await apiFetch( {
+						path: `/rest/v1.1/videos/${ media.videopress_guid }`,
+						addSiteSlug: false,
+					} );
+					attributes.src = `${ videoPressData.file_url_base.https }${
+						videoPressData.files.hd.mp4
+					}`;
+				}
+			}
+			this.props.setAttributes( attributes );
+		};
```

Comment:
I think we want to make sure we understand what we're doing here by changing `setAttributes()` to an `async` function and how that gets transformed. because I wasn't sure myself I did some testing.

the following sets up three functions: one is a simple synchronous function and the other two are asynchronous in some way. if the syntax is confusing at all just let me know and I can clarify what's going on

```js
const l = m => ( console.log( m ), m );

let a = 0, b = 0, c = 0;

const immediately = () => a + 1;
const afterwards = async () => b + 1;
const supposedly = () => {
  Promise.resolve().then( () => c += 1 );
  return c;
}

l( a ); l( immediately() ); // produces 0, 1
l( b ); l( afterwards() ); // produces 0, Promise { 1 }
l( c ); l( supposedly() ); // produces 0, 0
Promise.resolve().then( () => l( c ) ); // produces 1
```

`immediately()` is obvious and simple and also not in concern for this change - it's just a kind of control group

`afterwards()` returns a `Promise` because it's an `async` function. it doesn't return the simple number that it changes because of this wrapping. when I added side-effects inside of it they did apply immediately, so our call to `this.props.setAttributes( attributes )` should still have the intended effect, but there are two nuance changes here: first of all we no longer return the value from `setAttributes()` (which granted nothing probably uses and it may always be `undefined`); and second were to do that it would be wrapped in a `Promise` instead of being the value itself.

`supposedly()` is a way of writing `afterwards()` that leaves the return values intact. we can see that its side-effects are deferred because the `Promise` code runs _after_ all of the current function's code has run. it's not exactly equivalent to `afterwards()` because I couldn't do that and demonstrate the point. here's the way we'd rewrite `afterwards()` to match it.

```js
const afterwards = async () => {
	await Promise.resolve();
	return b+= 1;
}
```

with this refactor the side-effects here too are deferred and the function itself returns `Promise {pending}`

---

what does this do during our JavaScript compilation? here are those three files as translated by [Babel's REPL](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBANjAvDAtkgfDAFDUkRwCmAdHCAOY5oCUANKjNQNwBQLRsAhkjAAyscYAIx792hWMFGsWeaDACWKFIQAmCzlEJwAnjyzUMMbgGoYARhlyuAMy0AnAO6d7qiD04QdYKQaMBvFhhjZwVYAAV7EBQFCBJ7QggCADdCA1ZghKgAV3swYRgzSxYAXzZrGAhsgAdqkDjVXX1DRExA4Mjo2PjElLTqYigAC0IwHD9W3ELkc0YMmCzc_OBWMvYcbmZ1xWU1DS1dPy3tkS24DbtCJxc3I5lzqbOcKtr6tUPDLaA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=false&presets=es2015%2Cenv&prettier=true&targets=&version=6.26.0&envVersion=1.6.2)

```js
// const immediately = () => a + 1;

var immediately = function immediately() {
  return (a += 1);
};
```

```js
// const afterwards = async () => {
//   await Promise.resolve();
//   return b += 1;
// }

function _asyncToGenerator(fn) {
  return function() {
    var gen = fn.apply(this, arguments);
    return new Promise(function(resolve, reject) {
      function step(key, arg) {
        try {
          var info = gen[key](arg);
          var value = info.value;
        } catch (error) {
          reject(error);
          return;
        }
        if (info.done) {
          resolve(value);
        } else {
          return Promise.resolve(value).then(
            function(value) {
              step("next", value);
            },
            function(err) {
              step("throw", err);
            }
          );
        }
      }
      return step("next");
    });
  };
}

var afterwards = (function() {
  var _ref = _asyncToGenerator(
    /*#__PURE__*/ regeneratorRuntime.mark(function _callee() {
      return regeneratorRuntime.wrap(
        function _callee$(_context) {
          while (1) {
            switch ((_context.prev = _context.next)) {
              case 0:
                _context.next = 2;
                return Promise.resolve();

              case 2:
                return _context.abrupt("return", (b += 1));

              case 3:
              case "end":
                return _context.stop();
            }
          }
        },
        _callee,
        undefined
      );
    })
  );

  return function afterwards() {
    return _ref.apply(this, arguments);
  };
})();
```

```js
// const supposedly = () => {
//   Promise.resolve().then( () => c += 1 );
//   return c;
// }

var supposedly = function supposedly() {
  Promise.resolve().then(function() {
    return (c += 1);
  });
  return c;
};
```

one of these three stands out 😉 - I don't know the impact of the cost for this overhead, but I do know that it's not free and I do find it confusing especially when I think we can get by without it.

---

what am I saying in this comment?

 - I hope that, in case you didn't already know it, you learned something. I did! I wasn't clear-enough in my understanding to be able to say how the return values of those functions would behave.
 - I would like to propose a change in the type of function used: see the following…

```js
setAttributes = attributes => {
	const { id } = attributes;

	if (id) {
		apiFetch( { path: `/rest/v1.1/media/${ id }` } )
			.then( ( { videopress_guid: id } ) => id && apiFetch( { path: `/rest/v1.1/videos/${ id }`, addSiteSlug: false } ) )
			.then( data => data && this.props.setAttributes( { src: `${ data.file_url_base.https }${ data.files.hd.mp4 }` } ) );
	}

	return this.props.setAttributes(attributes);
};
```

There are a couple primary impacts here beyond switching from `async` to `Promise`
 - we leave the return value the way it was before. we only care about the additional side-effect here so we'll let that spin up in the event stack through the `Promise` and leave the original flow intact
 - we leave execution through this function synchronous even when the `id` is being set. we _don't_ set the `id` attribute synchronously, though it raises the question if we should
 - when setting the VideoPress URL we _only_ update the `URL` attribute instead of all of the attributes. we don't need to made the other attributes freeze in the meantime and we want to avoid a race condition where other attributes may have changed in the time it time it took to issue the receive the network calls. we also don't want to interrupt the normal block functioning in the case where these API calls fail for some reason - we have no error-handling in the API calls and honestly that's probably fine since the VideoPress URL is an _enhancement_.

In this state some important questions become clearer to me:
 - should we go ahead and set the original URL and then set the updated URL when a successful call-chain to VideoPress returns? if we don't set the original URL then we'll be introducing an inconsistent state into the UI during the time these two sequential network requests go out. they _have_ to be sequential since once relies on data from the other
 - what should happen if the network calls fail? should we try again? flag the editor? we're chaining requests which means we have multiplying ways they can fail.
 - should we modify the `/media` endpoint to accept a new field option to return the VideoPress URL? this would remove one chain in the network sequence and simplify the error handling. (this would be a good follow-up change)
 - what should we do _if_ the other attributes are changed during the time these network requests are in flight? should we cancel the VideoPress request? I want to avoid the race where the editor changes the video twice but because of how HTTP works we end up with the VideoPress URL for the first one. if we get to the final `then()` and the original id of the video isn't the same one from what we received then maybe we should cancel our update.

this is already way too long so I won't show the updated code to reflect some of these questions, needless to say it's a slight bit more complicated but a lotta bit more resilient.

Reply:
```js
setAttributes = attributes => {
	const { id } = attributes;

	if (id) {
		apiFetch( { path: `/rest/v1.1/media/${ id }` } )
			.then( ( { videopress_guid: guid } ) => {
				const { id: thisId } = this.props.attributes;

				if ( id !== thisId ) {
					throw new Error( 'changed_video_id' );
				}

				if ( ! guid ) {
					throw new Error( 'missing_videopress_guid' );
				}
				
				return apiFetch( { path: `/rest/v1.1/videos/${ guid }`, addSiteSlug: false } );
			} )
			.then( data => {
				const { id: thisId } = this.props.attributes;

				if ( id !== thisId ) {
					throw new Error( 'changed_video_id' );
				}

				if ( ! data ) {
					throw new Error( 'missing_videopress_data' );
				}

				this.props.setAttributes( { src: `${ data.file_url_base.https }${ data.files.hd.mp4 }` } );
			} )
			.catch( error => {
				switch ( error.message ) {
					case 'changed_video_id':
						return debug( 'video id was changed in the editor while fetching data for the previous video id');

					case 'missing_videopress_guid':
						return debug( `no VaultPress guid found in '/media/${ id }' response: aborting` );

					case 'missing_videopress_data':
						return debug( `no VaultPress video information found in '/videos/' endpoint for media id ${id}` );

					default:
						return debug( `Unexpected problem getting VideoPress data: ${ error.message }` );
				};
			} );
	}

	return this.props.setAttributes(attributes);
};
```

Reply:
Lovely! Thanks for the clear explanation. It totally makes sense so I applied some changes in 7eabd21 based on your suggestions: 
- `setAttributes` is not an async function anymore. Instead, so we leave the return value the way it was before but with one single change: all the attributes are set as normal except the src attribute. This is to avoid an extra download by the browser if we're going to finally replace the src with the VideoPress URL. If anything goes wrong, we set the original src attribute.
- We check now that the id attribute has not changed in order to avoid race conditions.
- We use the best resolution (hd > dvd > std) and format possible (mp4 > ogv). If anything is missing, we throw an error.